src/IO/S3/Credentials.cpp

```diff
@@ -838,6 +845,158 @@ S3CredentialsProviderChain::S3CredentialsProviderChain(
     AddProvider(std::make_shared<Aws::Auth::ProfileConfigFileAWSCredentialsProvider>());
 }
 
+AssumeRoleRequest::AssumeRoleRequest(std::string role_arn_, std::string role_session_name_)
+    : role_arn(std::move(role_arn_)), role_session_name(std::move(role_session_name_))
+{
+}
+
+Aws::Http::HeaderValueCollection AssumeRoleRequest::GetHeaders() const
+{
+    return {{Aws::Http::HeaderValuePair(Aws::Http::CONTENT_TYPE_HEADER, Aws::AMZN_XML_CONTENT_TYPE)}};
+}
+
+void AssumeRoleRequest::AddQueryStringParameters(Aws::Http::URI & uri) const
+{
+    uri.AddQueryStringParameter("RoleArn", role_arn);
+    uri.AddQueryStringParameter("RoleSessionName", role_session_name);
+}
+
+AssumeRoleResult::AssumeRoleResult(Aws::AmazonWebServiceResult<Aws::Utils::Xml::XmlDocument> result)
+{
+    using namespace Aws::Utils::Xml;
+    const auto & xml_document = result.GetPayload();
+    auto response_node = xml_document.GetRootElement();
+
+    if (response_node.IsNull())
+    {
+        LOG_WARNING(log, "Response is empty");
+        return;
+    }
+
+    auto assume_role_result_node = response_node.FirstChild("AssumeRoleResult");
+    if (assume_role_result_node.IsNull())
+    {
+        LOG_WARNING(log, "AssumeRoleResult node is missing");
+        return;
+    }
+
+    auto credentials_node = assume_role_result_node.FirstChild("Credentials");
+    if (credentials_node.IsNull())
+    {
+        LOG_WARNING(log, "Credentials node is missing");
+        return;
+    }
+
+    const auto get_credential_value_from_node = [&](auto & dest, const std::string & node_name)
+    {
+        if (auto node = credentials_node.FirstChild(node_name);
+            !node.IsNull())
+            dest = DecodeEscapedXmlText(node.GetText());
+        else
+            LOG_WARNING(log, "{} node is missing", node_name);
+    };
+
+    get_credential_value_from_node(access_key_id, "AccessKeyId");
+    get_credential_value_from_node(secret_access_key, "SecretAccessKey");
+    get_credential_value_from_node(session_token, "SessionToken");
+    get_credential_value_from_node(expiration, "Expiration");
+}
+
+AWSAssumeRoleClient::AWSAssumeRoleClient(
+    const std::shared_ptr<Aws::Auth::AWSCredentialsProvider> & credentials_provider,
+    const Aws::Client::ClientConfiguration & client_configuration,
+    const std::string & sts_endpoint_override)
+    : Aws::Client::AWSXMLClient(
+        client_configuration,
+        std::make_shared<Aws::Auth::DefaultAuthSignerProvider>(
+            credentials_provider,
+            "sts",
+            Aws::Region::ComputeSignerRegion(client_configuration.region),
+            Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy::Always,
+            /*doubleEncodeValue*/ false),
+        std::make_shared<Aws::Client::S3ErrorMarshaller>())
+{
+    if (!sts_endpoint_override.empty())
+    {
+        endpoint.SetURL(sts_endpoint_override);
```

Comment:
NOTE: This ends up doing a POST to this URL without the `Action` parameter.  You're only appending `?Action=AssumeRole&Version=2011-06-15` below on the non-overridden endpoint.

I think it's important that those query params get used regardless of override vs. not.  The params aren't technically part of the "endpoint" (which should just be the base URL imho).